### PR TITLE
Exp/serializable projectevaluation

### LINKED ISSUE #6260: Serializable MSBuild.Project support
Support for CPS solution load by caching project evaluation - https://dev.azure.com/devdiv/DevDiv/_wiki/wikis/DevDiv.wiki/20932/Project-evaluation-cache

### PR DESCRIPTION
**Pull request by jackcartman**
Fixes #6260

### Context
The evaluation result is being cached now on the CPS side to speed up solution load. For the cases where CPS has it's linked read-only Project, we would like to create a ProjectInstance as well from it, instead of triggering a evaluation for the sake of creating a ProjectInstance

### Changes Made
- Add a constructor to ProjectInstance that builds itself from the Project.
- Also exposed the existing GetAllGlobs override to ProjectLink. Previously never exposed since ProjectLink was used only for csproj that does not really care about globs.

### Testing
Added a test for the new constructor.

### CODE REVIEW

**Review by Forgind — APPROVED** [review]

File: src/Build/Instance/ProjectInstance.cs [hunk 1]
```diff
@@ -2797,34 +2855,30 @@ private void CreateGlobalPropertiesSnapshot(Evaluation.Project.Data data)
...
-                }
+                _itemsByEvaluatedInclude.Add(item.EvaluatedInclude, projectItemToInstanceMap[item]);
             }
         }
 
         /// <summary>
         /// Create Items snapshot
         /// </summary>
-        private Dictionary<ProjectItem, ProjectItemInstance> CreateItemsSnapshot(Evaluation.Project.Data data, bool keepEvaluationCache)
+        private Dictionary<ProjectItem, ProjectItemInstance> CreateItemsSnapshot(ICollection<ProjectItem> items, int itemTypecount, bool keepEvaluationCache)
```
Comment (ladipro) [inline src/Build/Instance/ProjectInstance.cs:2875]: nit: `itemTypeCount` (capital C).

File: src/Build/Instance/ProjectInstance.cs [hunk 2]
```diff
@@ -2852,7 +2906,12 @@ private void CreateEvaluatedIncludeSnapshotIfRequested(bool keepEvaluationCache,
                     }
                 }
 
-                ProjectItemInstance instance = new ProjectItemInstance(this, item.ItemType, ((IItem)item).EvaluatedIncludeEscaped, item.EvaluatedIncludeBeforeWildcardExpansionEscaped, directMetadata, inheritedItemDefinitions, item.Xml.ContainingProject.EscapedFullPath);
+                var evaluatedIncludeEscaped = ((IItem)item).EvaluatedIncludeEscaped;
+                evaluatedIncludeEscaped ??= item.EvaluatedInclude;
+                var evaluatedIncludeBeforeWildcardExpansionEscaped = item.EvaluatedIncludeBeforeWildcardExpansionEscaped;
+                evaluatedIncludeBeforeWildcardExpansionEscaped ??= item.EvaluatedInclude;
```
Comment (ladipro) [inline src/Build/Instance/ProjectInstance.cs:2912]: Can you please add a comment to these changes? It's not clear why the fallback to `item.EvaluatedInclude` is needed.

**Review by ladipro — COMMENTED** [review]
Looks good! I've left a nit and a question inline.

File: src/Build/Instance/ProjectInstance.cs [hunk 3]
```diff
@@ -326,6 +326,59 @@ public ProjectInstance(ProjectRootElement xml, IDictionary<string, string> globa
         {
         }
 
+        /// <summary>
+        /// Creates a ProjectInstance from an external created <see cref="Project"/>.
+        /// Properties and items are cloned immediately and only the instance data is stored.
+        /// </summary>
+        public ProjectInstance(Project project, ProjectInstanceSettings settings)
```
Comment (cdmihai) [inline src/Build/Instance/ProjectInstance.cs:333]: How is this different from the other [ProjectInstance(Evaluation.Project.Data)](https://github.com/dotnet/msbuild/blob/97463e3cdb3ce4ccf9638eb68fd9f82646ca6a11/src/Build/Instance/ProjectInstance.cs#L451) constructor which is called from [Project.CreateProjectInstance](https://github.com/dotnet/msbuild/blob/97463e3cdb3ce4ccf9638eb68fd9f82646ca6a11/src/Build/Definition/Project.cs#L1333)? They seem the same to me, so I'd rather not add this additional constructor (unless I'm missing something).

Comment (cdmihai) [inline src/Build/Instance/ProjectInstance.cs:333]: How is this different from the other [ProjectInstance(Evaluation.Project.Data)](https://github.com/dotnet/msbuild/blob/97463e3cdb3ce4ccf9638eb68fd9f82646ca6a11/src/Build/Instance/ProjectInstance.cs#L451) constructor which is called from [Project.CreateProjectInstance](https://github.com/dotnet/msbuild/blob/97463e3cdb3ce4ccf9638eb68fd9f82646ca6a11/src/Build/Definition/Project.cs#L1333)? They seem the same to me, so I'd rather not add this additional constructor (unless I'm missing something).

**Review by cdmihai — CHANGES_REQUESTED** [review]

File: src/Build/Instance/ProjectInstance.cs (hunk 2 above)
Comment (arkalyanms) [reply src/Build/Instance/ProjectInstance.cs:2913]: Added. It's for external ProjectLink based support.

File: src/Build/Instance/ProjectInstance.cs (hunk 3 above)
Comment (cdmihai) [reply src/Build/Instance/ProjectInstance.cs:333]: Ah, okay, let me rephrase it to ensure that I got it right: your scenario is to create a `ProjectInstance` object residing in Process_1 based on a `Project` object residing in Process_2. You want to avoid marshaling the `Project.Data` object from Process_2 to Process_1 and instead have the `ProjectInstance` constructor from Process_1 do remote calls for each `Project` data field (one "rpc" call for properties, one for items, one for targets, etc).

Yea, in this case reading the data directly from the remote Project object is more straighforward than manually coding up a stub/proxy for `Project.Data` as well. Assuming the two processes always reside on the same machine, otherwise the rpc overhead might make it worth it to serialize the entire Project.Data over one rpc call.

However, I'd just delete the old `ProjectInstance(Evaluation.Project.Data)` constructor. It's just code duplication with your new constructor.

Comment (arkalyanms) [reply src/Build/Instance/ProjectInstance.cs:333]: Yes, that's a good summary. The old `ProjectInstance(Evaluation.Project.Data)` constructor still will get used as-is in the non cached mode in VS where MSBuild gets called for actual evaluation. I would like to keep it that way for feature isolation reasons, atleast until the cache mode feature which will initially go in behind a feature flag gets more test coverage.

**Review by cdmihai — APPROVED** [review]
